plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,539 @@
+import org.opensearch.threadpool.ThreadPool;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/** Manager class for ML models and nodes. It contains ML model auto redeploy operations etc. */
+@Log4j2
+public class MLModelAutoReDeployer {
+
+    private final ClusterService clusterService;
+    private final Client client;
+    private final ThreadPool threadPool;
+    private final NodeEnvironment nodeEnvironment;
+    private final Settings settings;
+    private final NamedXContentRegistry xContentRegistry;
+    private volatile Boolean enableAutoReDeployModel;
+    private volatile Boolean onlyRunOnMlNode;
+    private volatile Integer autoDeployMaxRetryTimes;
+    private volatile boolean allowCustomDeploymentPlan;
+    private volatile boolean isMasterNode;
+
+    /**
+     * constructor method， init all the params necessary for model auto redeploying
+     *
+     * @param clusterService clusterService
+     * @param threadPool threadPool
+     * @param client client
+     * @param xContentRegistry xContentRegistry
+     * @param settings settings
+     * @param nodeEnvironment nodeEnvironment
+     */
+    public MLModelAutoReDeployer(
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        Client client,
+        Settings settings,
+        NodeEnvironment nodeEnvironment,
+        NamedXContentRegistry xContentRegistry
+    ) {
+        this.clusterService = clusterService;
+        this.client = client;
+        this.threadPool = threadPool;
+        this.settings = settings;
+        this.nodeEnvironment = nodeEnvironment;
+        this.xContentRegistry = xContentRegistry;
+
+        enableAutoReDeployModel = ML_COMMONS_MODEL_AUTO_REDEPLOY_ENABLE.get(settings);
+        onlyRunOnMlNode = ML_COMMONS_ONLY_RUN_ON_ML_NODE.get(settings);
+        autoDeployMaxRetryTimes = ML_COMMONS_MODEL_AUTO_REDEPLOY_RETRY_TIMES.get(settings);
+        allowCustomDeploymentPlan = ML_COMMONS_ALLOW_CUSTOM_DEPLOYMENT_PLAN.get(settings);
+
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_COMMONS_MODEL_AUTO_REDEPLOY_ENABLE, it -> enableAutoReDeployModel = it);
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ML_COMMONS_ONLY_RUN_ON_ML_NODE, it -> onlyRunOnMlNode = it);
+
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_COMMONS_MODEL_AUTO_REDEPLOY_RETRY_TIMES, it -> autoDeployMaxRetryTimes = it);
+
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_COMMONS_ALLOW_CUSTOM_DEPLOYMENT_PLAN, it -> allowCustomDeploymentPlan = it);
+    }
+
+    /** the main method: model auto redeploying */
+    public void autoReDeployModel() {
+        log.info("auto redeploy model enabled: {} ", enableAutoReDeployModel);
+
+        // if we don't need to redeploy automatically, just return without doing anything
+        if (!enableAutoReDeployModel) {
```

Comment:
Move the master node check here to make sure all the checks are gathered together, this will make it easy for  reading code.

Reply:
Because at here, the whole cluster state has not been initialized, and I use the `clusterservice` class to check whether the node is master node, at this time `clusterService.state().nodes()` maybe throw `nullPointerException`, so we can not do master node check at here